[ICE]: `Unexpected type for constructor`

<!--
[31mICE[0m: Rustc ./a.rs '-Znext-solver=globally -ooutputfile -Zdump-mir-dir=dir' 'error: internal compiler error: compiler/rustc_pattern_analysis/src/rustc.rs:260:22: Unexpected type for constructor `Struct`: Alias(Yes, Alias { kind: Opaque { def_id: DefId(0:8 ~ a[7aa2]::main::U::{opaque#0}) }, args: [], .. })', 'error: internal compiler error: compiler/rustc_pattern_analysis/src/rustc.rs:260:22: Unexpected type for constructor `Struct`: Alias(Yes, Alias { kind: Opaque { def_id: DefId(0:8 ~ a[7aa2]::main::U::{opaque#0}) }, args: [], .. })'
File: /tmp/im/11/a.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -Znext-solver=globally
fn main() {
    struct Foo(U);
    type U = impl Copy;
    let foo: _ = Foo(());
    let Foo(()) = foo;
}

````

original:
````rust
#![feature(type_alias_impl_trait)]
fn main() {
    struct Foo(U);
    type U = impl Copy;
    let foo: _ = Foo(());
    let Foo(()) = foo;
}

````

Version information
````
rustc 1.98.0-nightly (40557f622 2026-06-25)
binary: rustc
commit-hash: 40557f6225e337d68c8d4f086557ce54135f5dd9
commit-date: 2026-06-25
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.8

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/40557f6225e337d68c8d4f086557ce54135f5dd9/compiler/rustc_pattern_analysis/src/rustc.rs#L254-L266

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Znext-solver=globally`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: `impl Trait` in type aliases is unstable
 --> /tmp/icemaker_global_tempdir.Q1eSBCVKywuU/rustc_testrunner_tmpdir_reporting.oAi1oVF0KRX0/mvce.rs:3:14
  |
3 |     type U = impl Copy;
  |              ^^^^^^^^^
  |
  = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
  = help: add `#![feature(type_alias_impl_trait)]` to the crate attributes to enable
  = note: this compiler was built on 2026-06-25; consider upgrading it if it is out of date

error: internal compiler error: /rustc-dev/40557f6225e337d68c8d4f086557ce54135f5dd9/compiler/rustc_pattern_analysis/src/rustc.rs:260:22: Unexpected type for constructor `Struct`: Alias(Yes, Alias { kind: Opaque { def_id: DefId(0:8 ~ mvce[2093]::main::U::{opaque#0}) }, args: [], .. })


thread 'rustc' (125234) panicked at /rustc-dev/40557f6225e337d68c8d4f086557ce54135f5dd9/compiler/rustc_pattern_analysis/src/rustc.rs:260:22:
Box<dyn Any>
stack backtrace:
   0:     0x7fa1124a2f49 - <<std[2a68cadfb35bda34]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[2ec008d822edb004]::fmt::Display>::fmt
   1:     0x7fa112c0ec08 - core[2ec008d822edb004]::fmt::write
   2:     0x7fa1124b95b6 - <std[2a68cadfb35bda34]::sys::stdio::unix::Stderr as std[2a68cadfb35bda34]::io::Write>::write_fmt
   3:     0x7fa1124780ae - std[2a68cadfb35bda34]::panicking::default_hook::{closure#0}
   4:     0x7fa112495ac3 - std[2a68cadfb35bda34]::panicking::default_hook
   5:     0x7fa1114a6771 - std[2a68cadfb35bda34]::panicking::update_hook::<alloc[9c08d3116dc5205f]::boxed::Box<rustc_driver_impl[b066d2030f6db219]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7fa112495f12 - std[2a68cadfb35bda34]::panicking::panic_with_hook
   7:     0x7fa1114d4591 - std[2a68cadfb35bda34]::panicking::begin_panic::<rustc_errors[44fc769b9dccdf7a]::ExplicitBug>::{closure#0}
   8:     0x7fa1114cd306 - std[2a68cadfb35bda34]::sys::backtrace::__rust_end_short_backtrace::<std[2a68cadfb35bda34]::panicking::begin_panic<rustc_errors[44fc769b9dccdf7a]::ExplicitBug>::{closure#0}, !>
   9:     0x7fa1114cb297 - std[2a68cadfb35bda34]::panicking::begin_panic::<rustc_errors[44fc769b9dccdf7a]::ExplicitBug>
  10:     0x7fa1114df721 - <rustc_errors[44fc769b9dccdf7a]::diagnostic::BugAbort as rustc_errors[44fc769b9dccdf7a]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  11:     0x7fa111a43689 - rustc_middle[b4a081e030a285a9]::util::bug::opt_span_bug_fmt::<rustc_span[c2b6ba5c55640e04]::span_encoding::Span>::{closure#0}
  12:     0x7fa111a437f2 - rustc_middle[b4a081e030a285a9]::ty::context::tls::with_opt::<rustc_middle[b4a081e030a285a9]::util::bug::opt_span_bug_fmt<rustc_span[c2b6ba5c55640e04]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  13:     0x7fa111a2f41b - rustc_middle[b4a081e030a285a9]::ty::context::tls::with_context_opt::<rustc_middle[b4a081e030a285a9]::ty::context::tls::with_opt<rustc_middle[b4a081e030a285a9]::util::bug::opt_span_bug_fmt<rustc_span[c2b6ba5c55640e04]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  14:     0x7fa10ed1f674 - rustc_middle[b4a081e030a285a9]::util::bug::bug_fmt
  15:     0x7fa112f24989 - rustc_pattern_analysis[1928fcd5bf71bb97]::usefulness::compute_exhaustiveness_and_usefulness::<rustc_pattern_analysis[1928fcd5bf71bb97]::rustc::RustcPatCtxt>::{closure#0}
  16:     0x7fa112f1fb78 - rustc_pattern_analysis[1928fcd5bf71bb97]::usefulness::compute_exhaustiveness_and_usefulness::<rustc_pattern_analysis[1928fcd5bf71bb97]::rustc::RustcPatCtxt>::{closure#0}
  17:     0x7fa112f17e08 - rustc_pattern_analysis[1928fcd5bf71bb97]::rustc::analyze_match
  18:     0x7fa113c57065 - <rustc_mir_build[71e6f7ecde5855b7]::thir::pattern::check_match::MatchVisitor>::check_binding_is_irrefutable
  19:     0x7fa112c135c7 - <rustc_mir_build[71e6f7ecde5855b7]::thir::pattern::check_match::MatchVisitor as rustc_middle[b4a081e030a285a9]::thir::visit::Visitor>::visit_expr
  20:     0x7fa112c12fb2 - <rustc_mir_build[71e6f7ecde5855b7]::thir::pattern::check_match::MatchVisitor as rustc_middle[b4a081e030a285a9]::thir::visit::Visitor>::visit_expr
  21:     0x7fa112f13a10 - rustc_mir_build[71e6f7ecde5855b7]::thir::pattern::check_match::check_match
  22:     0x7fa112f136c7 - rustc_query_impl[7ecdd8637321aee]::query_impl::check_match::invoke_provider_fn::__rust_begin_short_backtrace
  23:     0x7fa113ab1371 - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_data_structures[642e38e42f99d6a5]::vec_cache::VecCache<rustc_span[c2b6ba5c55640e04]::def_id::LocalDefId, rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 1usize]>, rustc_middle[b4a081e030a285a9]::dep_graph::graph::DepNodeIndex>, false>
  24:     0x7fa113ab1076 - rustc_query_impl[7ecdd8637321aee]::query_impl::check_match::execute_query_non_incr::__rust_end_short_backtrace
  25:     0x7fa1136351d5 - rustc_mir_build[71e6f7ecde5855b7]::builder::build_mir_inner_impl
  26:     0x7fa112c0f8aa - rustc_mir_transform[29a06d6b7ef68a2]::mir_built
  27:     0x7fa112ff2e69 - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_data_structures[642e38e42f99d6a5]::vec_cache::VecCache<rustc_span[c2b6ba5c55640e04]::def_id::LocalDefId, rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[b4a081e030a285a9]::dep_graph::graph::DepNodeIndex>, false>
  28:     0x7fa112ff2b03 - rustc_query_impl[7ecdd8637321aee]::query_impl::mir_built::execute_query_non_incr::__rust_end_short_backtrace
  29:     0x7fa113ab235d - rustc_mir_transform[29a06d6b7ef68a2]::ffi_unwind_calls::has_ffi_unwind_calls
  30:     0x7fa113ab1da3 - rustc_query_impl[7ecdd8637321aee]::query_impl::has_ffi_unwind_calls::invoke_provider_fn::__rust_begin_short_backtrace
  31:     0x7fa113ab1371 - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_data_structures[642e38e42f99d6a5]::vec_cache::VecCache<rustc_span[c2b6ba5c55640e04]::def_id::LocalDefId, rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 1usize]>, rustc_middle[b4a081e030a285a9]::dep_graph::graph::DepNodeIndex>, false>
  32:     0x7fa113ab0f42 - rustc_query_impl[7ecdd8637321aee]::query_impl::has_ffi_unwind_calls::execute_query_non_incr::__rust_end_short_backtrace
  33:     0x7fa1137b82ca - rustc_mir_transform[29a06d6b7ef68a2]::mir_promoted
  34:     0x7fa1137b812e - rustc_query_impl[7ecdd8637321aee]::query_impl::mir_promoted::invoke_provider_fn::__rust_begin_short_backtrace
  35:     0x7fa1131e3e01 - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_data_structures[642e38e42f99d6a5]::vec_cache::VecCache<rustc_span[c2b6ba5c55640e04]::def_id::LocalDefId, rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 16usize]>, rustc_middle[b4a081e030a285a9]::dep_graph::graph::DepNodeIndex>, false>
  36:     0x7fa113a94307 - rustc_query_impl[7ecdd8637321aee]::query_impl::mir_promoted::execute_query_non_incr::__rust_end_short_backtrace
  37:     0x7fa113a944e9 - rustc_borrowck[cb69b73e2bece8e3]::mir_borrowck
  38:     0x7fa113a943e5 - rustc_query_impl[7ecdd8637321aee]::query_impl::mir_borrowck::invoke_provider_fn::__rust_begin_short_backtrace
  39:     0x7fa112ff2e69 - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_data_structures[642e38e42f99d6a5]::vec_cache::VecCache<rustc_span[c2b6ba5c55640e04]::def_id::LocalDefId, rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[b4a081e030a285a9]::dep_graph::graph::DepNodeIndex>, false>
  40:     0x7fa112fefe91 - rustc_query_impl[7ecdd8637321aee]::query_impl::mir_borrowck::execute_query_non_incr::__rust_end_short_backtrace
  41:     0x7fa11162e992 - <rustc_hir_analysis[bef09c3ec5a10b3a]::collect::type_of::opaque::TaitConstraintLocator>::check
  42:     0x7fa11166d351 - rustc_hir_analysis[bef09c3ec5a10b3a]::collect::type_of::opaque::find_opaque_ty_constraints_for_tait
  43:     0x7fa11499676d - rustc_hir_analysis[bef09c3ec5a10b3a]::collect::type_of::type_of_opaque.cold
  44:     0x7fa112c2bf8d - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_middle[b4a081e030a285a9]::query::caches::DefIdCache<rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 8usize]>>, false>
  45:     0x7fa113d15a5a - rustc_query_impl[7ecdd8637321aee]::query_impl::type_of_opaque::execute_query_non_incr::__rust_end_short_backtrace
  46:     0x7fa112f543bd - rustc_hir_analysis[bef09c3ec5a10b3a]::collect::type_of::type_of
  47:     0x7fa112f47c66 - rustc_query_impl[7ecdd8637321aee]::query_impl::type_of::invoke_provider_fn::__rust_begin_short_backtrace
  48:     0x7fa112c2bf8d - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_middle[b4a081e030a285a9]::query::caches::DefIdCache<rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 8usize]>>, false>
  49:     0x7fa112c2bb10 - rustc_query_impl[7ecdd8637321aee]::query_impl::type_of::execute_query_non_incr::__rust_end_short_backtrace
  50:     0x7fa1139f8c0f - rustc_hir_analysis[bef09c3ec5a10b3a]::check::check::check_opaque
  51:     0x7fa113ad4909 - rustc_hir_analysis[bef09c3ec5a10b3a]::check::check::check_item_type
  52:     0x7fa113ab6226 - rustc_hir_analysis[bef09c3ec5a10b3a]::check::wfcheck::check_well_formed
  53:     0x7fa113ab6207 - rustc_query_impl[7ecdd8637321aee]::query_impl::check_well_formed::invoke_provider_fn::__rust_begin_short_backtrace
  54:     0x7fa113ab1371 - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_data_structures[642e38e42f99d6a5]::vec_cache::VecCache<rustc_span[c2b6ba5c55640e04]::def_id::LocalDefId, rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 1usize]>, rustc_middle[b4a081e030a285a9]::dep_graph::graph::DepNodeIndex>, false>
  55:     0x7fa113ab1107 - rustc_query_impl[7ecdd8637321aee]::query_impl::check_well_formed::execute_query_non_incr::__rust_end_short_backtrace
  56:     0x7fa113ab32ee - rustc_hir_analysis[bef09c3ec5a10b3a]::check::wfcheck::check_type_wf
  57:     0x7fa113ab2c55 - rustc_query_impl[7ecdd8637321aee]::query_impl::check_type_wf::invoke_provider_fn::__rust_begin_short_backtrace
  58:     0x7fa113d0bac6 - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_middle[b4a081e030a285a9]::query::caches::SingleCache<rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 1usize]>>, false>
  59:     0x7fa113d0b8a9 - rustc_query_impl[7ecdd8637321aee]::query_impl::check_type_wf::execute_query_non_incr::__rust_end_short_backtrace
  60:     0x7fa112feca5a - rustc_hir_analysis[bef09c3ec5a10b3a]::check_crate
  61:     0x7fa112feef74 - rustc_interface[daa44c320b53a0f8]::passes::analysis
  62:     0x7fa113d0c8c7 - rustc_query_impl[7ecdd8637321aee]::execution::try_execute_query::<rustc_middle[b4a081e030a285a9]::query::caches::SingleCache<rustc_middle[b4a081e030a285a9]::query::erase::ErasedData<[u8; 0usize]>>, false>
  63:     0x7fa113d0c5a9 - rustc_query_impl[7ecdd8637321aee]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  64:     0x7fa113d55c00 - rustc_interface[daa44c320b53a0f8]::interface::run_compiler::<(), rustc_driver_impl[b066d2030f6db219]::run_compiler::{closure#0}>::{closure#1}
  65:     0x7fa113d232fa - std[2a68cadfb35bda34]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[daa44c320b53a0f8]::util::run_in_thread_with_globals<rustc_interface[daa44c320b53a0f8]::util::run_in_thread_pool_with_globals<rustc_interface[daa44c320b53a0f8]::interface::run_compiler<(), rustc_driver_impl[b066d2030f6db219]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  66:     0x7fa113d230ad - <std[2a68cadfb35bda34]::thread::lifecycle::spawn_unchecked<rustc_interface[daa44c320b53a0f8]::util::run_in_thread_with_globals<rustc_interface[daa44c320b53a0f8]::util::run_in_thread_pool_with_globals<rustc_interface[daa44c320b53a0f8]::interface::run_compiler<(), rustc_driver_impl[b066d2030f6db219]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[2ec008d822edb004]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  67:     0x7fa113d19550 - <std[2a68cadfb35bda34]::sys::thread::unix::Thread>::new::thread_start
  68:     0x7fa10d6981b9 - <unknown>
  69:     0x7fa10d71d21c - <unknown>
  70:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.98.0-nightly (40557f622 2026-06-25) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z next-solver=globally -Z dump-mir-dir=dir

query stack during panic:
#0 [check_match] match-checking `main`
#1 [mir_built] building MIR for `main`
#2 [has_ffi_unwind_calls] checking if `main` contains FFI-unwind calls
#3 [mir_promoted] promoting constants in MIR for `main`
#4 [mir_borrowck] borrow-checking `main`
#5 [type_of_opaque] computing type of opaque `main::U::{opaque#0}`
#6 [type_of] computing type of `main::U::{opaque#0}`
#7 [check_well_formed] checking that `main::U::{opaque#0}` is well-formed
#8 [check_type_wf] checking that types are well-formed
#9 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.

```

</p>
</details>

<!--
query stack:
#0 [check_match] match-checking `main`
#1 [mir_built] building MIR for `main`
#2 [has_ffi_unwind_calls] checking if `main` contains FFI-unwind calls
#3 [mir_promoted] promoting constants in MIR for `main`
#4 [mir_borrowck] borrow-checking `main`
#5 [type_of_opaque] computing type of opaque `main::U::{opaque#0}`
#6 [type_of] computing type of `main::U::{opaque#0}`
#7 [check_well_formed] checking that `main::U::{opaque#0}` is well-formed
#8 [check_type_wf] checking that types are well-formed
#9 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-type_alias_impl_trait +WG-trait-system-refactor

--- a/compiler/rustc_middle/src/ty/mod.rs
+++ b/compiler/rustc_middle/src/ty/mod.rs
@@ -1127,6 +1127,23 @@ impl<'tcx> TypingEnv<'tcx> {
         Self::new(tcx.param_env(def_id), TypingMode::non_body_analysis())
     }
 
+    /// Ideally we just use `TypingMode::PostTypeckUntilBorrowck`.
+    /// But that's not compatible with the old solver yet.
+    ///
+    /// FIXME: this should not be needed in the long term.
+    pub fn post_typeck_until_borrowck_for_mir_build(
+        tcx: TyCtxt<'tcx>,
+        def_id: LocalDefId,
+    ) -> TypingEnv<'tcx> {
+        if tcx.use_typing_mode_post_typeck_until_borrowck() {
+            TypingEnv::new(tcx.param_env(def_id.to_def_id()), ty::TypingMode::borrowck(tcx, def_id))
+        } else {
+            // FIXME(#132279): We're in a body, we should use a typing
+            // mode which reveals the opaque types defined by that body.
+            TypingEnv::non_body_analysis(tcx, def_id)
+        }
+    }
+
     pub fn post_analysis(tcx: TyCtxt<'tcx>, def_id: impl IntoQueryKey<DefId>) -> TypingEnv<'tcx> {
         TypingEnv::new(tcx.param_env_normalized_for_post_analysis(def_id), TypingMode::PostAnalysis)
     }

--- a/compiler/rustc_mir_build/src/builder/mod.rs
+++ b/compiler/rustc_mir_build/src/builder/mod.rs
@@ -505,9 +505,15 @@ fn construct_fn<'tcx>(
         );
     }
 
-    // FIXME(#132279): This should be able to reveal opaque
-    // types defined during HIR typeck.
-    let infcx = tcx.infer_ctxt().build(TypingMode::non_body_analysis());
+    let typing_mode = if tcx.use_typing_mode_post_typeck_until_borrowck() {
+        TypingMode::borrowck(tcx, fn_def)
+    } else {
+        // FIXME(#132279): This should be able to reveal opaque
+        // types defined during HIR typeck.
+        TypingMode::non_body_analysis()
+    };
+
+    let infcx = tcx.infer_ctxt().build(typing_mode);
     let mut builder = Builder::new(
         thir,
         infcx,
@@ -586,9 +592,15 @@ fn construct_const<'a, 'tcx>(
         _ => span_bug!(tcx.def_span(def), "can't build MIR for {:?}", def),
     };
 
-    // FIXME(#132279): We likely want to be able to use the hidden types of
-    // opaques used by this function here.
-    let infcx = tcx.infer_ctxt().build(TypingMode::non_body_analysis());
+    let typing_mode = if tcx.use_typing_mode_post_typeck_until_borrowck() {
+        TypingMode::borrowck(tcx, def)
+    } else {
+        // FIXME(#132279): This should be able to reveal opaque
+        // types defined during HIR typeck.
+        TypingMode::non_body_analysis()
+    };
+
+    let infcx = tcx.infer_ctxt().build(typing_mode);
     let mut builder =
         Builder::new(thir, infcx, def, hir_id, span, 0, const_ty, const_ty_span, None);
 

--- a/compiler/rustc_mir_build/src/builder/scope.rs
+++ b/compiler/rustc_mir_build/src/builder/scope.rs
@@ -969,8 +969,10 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             tcx: self.tcx,
             typeck_results,
             module: self.tcx.parent_module(self.hir_id).to_def_id(),
-            // FIXME(#132279): We're in a body, should handle opaques.
-            typing_env: rustc_middle::ty::TypingEnv::non_body_analysis(self.tcx, self.def_id),
+            typing_env: ty::TypingEnv::post_typeck_until_borrowck_for_mir_build(
+                self.tcx,
+                self.def_id,
+            ),
             dropless_arena: &dropless_arena,
             match_lint_level: self.hir_id,
             whole_match_span: Some(rustc_span::Span::default()),

--- a/compiler/rustc_mir_build/src/check_tail_calls.rs
+++ b/compiler/rustc_mir_build/src/check_tail_calls.rs
@@ -27,8 +27,7 @@ pub(crate) fn check_tail_calls(tcx: TyCtxt<'_>, def: LocalDefId) -> Result<(), E
         tcx,
         thir,
         found_errors: Ok(()),
-        // FIXME(#132279): we're clearly in a body here.
-        typing_env: ty::TypingEnv::non_body_analysis(tcx, def),
+        typing_env: ty::TypingEnv::post_typeck_until_borrowck_for_mir_build(tcx, def),
         is_closure,
         caller_def_id: def,
     };

--- a/compiler/rustc_mir_build/src/check_unsafety.rs
+++ b/compiler/rustc_mir_build/src/check_unsafety.rs
@@ -1093,8 +1093,7 @@ pub(crate) fn check_unsafety(tcx: TyCtxt<'_>, def: LocalDefId) {
         body_target_features,
         assignment_info: None,
         in_union_destructure: false,
-        // FIXME(#132279): we're clearly in a body here.
-        typing_env: ty::TypingEnv::non_body_analysis(tcx, def),
+        typing_env: ty::TypingEnv::post_typeck_until_borrowck_for_mir_build(tcx, def),
         inside_adt: false,
         warnings: &mut warnings,
         suggest_unsafe_block: true,

--- a/compiler/rustc_mir_build/src/thir/cx/mod.rs
+++ b/compiler/rustc_mir_build/src/thir/cx/mod.rs
@@ -99,9 +99,7 @@ impl<'tcx> ThirBuildCx<'tcx> {
         Self {
             tcx,
             thir: Thir::new(body_type),
-            // FIXME(#132279): We're in a body, we should use a typing
-            // mode which reveals the opaque types defined by that body.
-            typing_env: ty::TypingEnv::non_body_analysis(tcx, def),
+            typing_env: ty::TypingEnv::post_typeck_until_borrowck_for_mir_build(tcx, def),
             typeck_results,
             body_owner: def.to_def_id(),
             apply_adjustments: !find_attr!(tcx, hir_id, CustomMir(..)),

--- a/compiler/rustc_mir_build/src/thir/pattern/check_match.rs
+++ b/compiler/rustc_mir_build/src/thir/pattern/check_match.rs
@@ -39,8 +39,7 @@ pub(crate) fn check_match(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Result<(), Err
         tcx,
         thir: &*thir,
         typeck_results,
-        // FIXME(#132279): We're in a body, should handle opaques.
-        typing_env: ty::TypingEnv::non_body_analysis(tcx, def_id),
+        typing_env: ty::TypingEnv::post_typeck_until_borrowck_for_mir_build(tcx, def_id),
         hir_source: tcx.local_def_id_to_hir_id(def_id),
         let_source: LetSource::None,
         pattern_arena: &pattern_arena,

--- a/tests/ui/traits/next-solver/opaques/wrong-typing-mode-ice-1.rs
+++ b/tests/ui/traits/next-solver/opaques/wrong-typing-mode-ice-1.rs
@@ -1,0 +1,27 @@
+//@ compile-flags: -Znext-solver
+//@ edition: 2024
+
+// Previously, when building MIR body, we were using typing env
+// `non_body_analysis` which is wrong since we're indeed in a body.
+// It caused opaque types in defining body to be not revealed.
+// This caused opaque types in the defining body not to be revealed.
+
+#![feature(type_alias_impl_trait)]
+
+struct Task<F>(F);
+
+impl Task<Foo> {
+    fn spawn(&self, _: impl FnOnce() -> Foo) {}
+}
+type Foo = impl Sized;
+
+#[define_opaque(Foo)]
+fn foo() {
+    async fn cb() {}
+    Task::spawn(&POOL, cb)
+}
+
+static POOL: Task<Foo> = todo!();
+//~^ ERROR: evaluation panicked
+
+fn main() {}

--- a/tests/ui/traits/next-solver/opaques/wrong-typing-mode-ice-1.stderr
+++ b/tests/ui/traits/next-solver/opaques/wrong-typing-mode-ice-1.stderr
@@ -1,0 +1,9 @@
+error[E0080]: evaluation panicked: not yet implemented
+  --> $DIR/wrong-typing-mode-ice-1.rs:24:26
+   |
+LL | static POOL: Task<Foo> = todo!();
+   |                          ^^^^^^^ evaluation of `POOL` failed here
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0080`.

--- a/tests/ui/traits/next-solver/opaques/wrong-typing-mode-ice-2.rs
+++ b/tests/ui/traits/next-solver/opaques/wrong-typing-mode-ice-2.rs
@@ -1,0 +1,15 @@
+//@ compile-flags: -Znext-solver
+//@ check-pass
+
+// Previously, when building MIR body, we were using typing env
+// `non_body_analysis` which is wrong since we're indeed in a body.
+// It caused opaque types in defining body to be not revealed.
+// This caused opaque types in the defining body not to be revealed.
+
+#![feature(type_alias_impl_trait)]
+fn main() {
+    struct Foo(U);
+    type U = impl Copy;
+    let foo: _ = Foo(());
+    let Foo(()) = foo;
+}

